Implementation has two cases:
1. Node runs normally while SIGTERM is received
    => the SIGINT sent to node's pid
2. Node has been closed due to some issue (launch process still hanging around)
    => the SIGINT sent to pid 1

example:
Case 1)
```
[rplidar-1] [INFO] [1667544805.101603409] [jari.rplidar]: -------------- Loading parameters --------------
[static_transform_publisher-3] [INFO] [1667544805.102598479] [jari.fcu_to_garmin_static_transform_publisher]: Spinning until killed publishing transform from 'jari/fcu' to 'jari/garmin'
[rplidar-1] [INFO] [1667544805.104132697] [jari.rplidar]: RPLIDAR running on ROS 2 package rplidar_ros. SDK Version: '1.10.0'
Send SIGINT to pid 62
```

Case 2)
```
[static_transform_publisher-3] [INFO] [1667544712.393110939] [jari.fcu_to_garmin_static_transform_publisher]: Spinning until killed publishing transform from 'jari/fcu' to 'jari/garmin'
[ERROR] [rplidar-1]: process has died [pid 62, exit code -6, cmd '/opt/ros/galactic/lib/rplidar_ros2/rplidar --ros-args -r __node:=rplidar -r __ns:=/jari --params-file /opt/ros/galactic/share/rplidar_ros2/config/params.yaml --params-file /tmp/launch_params_bvkxvski -r topic_filtered_out:=~/scan_filtered -r topic_raw_out:=~/scan'].
Pattern not found, send SIGINT to pid 1
```

--- a/Dockerfile
+++ b/Dockerfile
@@ -15,8 +15,9 @@ FROM ghcr.io/tiiuae/fog-ros-baseimage:sha-3dcb78d
 # launch file checks env variables SIMULATION and DRONE_AIRFRAME
 # SIMULATION is by default 0. However, it can be set to 1
 # DRONE_AIRFRAME is by default "t-drone". However, it can be set to "holybro"
-ENTRYPOINT exec ros-with-env ros2 launch rplidar_ros2 sensors_launch.py
+ENTRYPOINT /entrypoint.sh
 
+COPY entrypoint.sh /entrypoint.sh
 COPY --from=builder /main_ws/ros-*-rplidar-ros2_*_amd64.deb /rplidar.deb
 
 # need update because ROS people have a habit of removing old packages pretty fast

--- a/entrypoint.sh
+++ b/entrypoint.sh
@@ -2,6 +2,30 @@
 
 _term() {
 	# FILL UP PROCESS SEARCH PATTERN HERE TO FIND PROPER PROCESS FOR SIGINT:
+	pattern="fcu_to_garmin_static_transform_publisher"
+
+	pid_value="$(ps -ax | grep $pattern | grep -v grep | awk '{ print $1 }')"
+	if [ "$pid_value" != "" ]; then
+		pid=$pid_value
+		echo "Send SIGINT to pid $pid"
+	else
+		pid=1
+		echo "Pattern not found, send SIGINT to pid $pid"
+	fi
+	kill -s SIGINT $pid
+
+	pattern="fcu_to_rplidar_static_transform_publisher"
+
+	pid_value="$(ps -ax | grep $pattern | grep -v grep | awk '{ print $1 }')"
+	if [ "$pid_value" != "" ]; then
+		pid=$pid_value
+		echo "Send SIGINT to pid $pid"
+	else
+		pid=1
+		echo "Pattern not found, send SIGINT to pid $pid"
+	fi
+	kill -s SIGINT $pid
+
 	pattern="rplidar_ros2/rplidar"
 
 	pid_value="$(ps -ax | grep $pattern | grep -v grep | awk '{ print $1 }')"

--- a/entrypoint.sh
+++ b/entrypoint.sh
@@ -1,0 +1,32 @@
+#!/bin/bash -e
+
+_term() {
+	# FILL UP PROCESS SEARCH PATTERN HERE TO FIND PROPER PROCESS FOR SIGINT:
+	pattern="rplidar_ros2/rplidar"
+
+	pid_value="$(ps -ax | grep $pattern | grep -v grep | awk '{ print $1 }')"
+	if [ "$pid_value" != "" ]; then
+		pid=$pid_value
+		echo "Send SIGINT to pid $pid"
+	else
+		pid=1
+		echo "Pattern not found, send SIGINT to pid $pid"
+	fi
+	kill -s SIGINT $pid
+}
+trap _term SIGTERM
+
+ros-with-env ros2 launch rplidar_ros2 sensors_launch.py &
+child=$!
+
+echo "Waiting for pid $child"
+wait $child
+RESULT=$?
+
+if [ $RESULT -ne 0 ]; then
+		echo "ERROR: RPLidar failed with code $RESULT" >&2
+		exit $RESULT
+else
+		echo "INFO: RPLidar finished successfully, but returning 125 code for docker to restart properly." >&2
+		exit 125
+fi

--- a/entrypoint.sh
+++ b/entrypoint.sh
@@ -14,12 +14,22 @@ _term() {
 	fi
 	kill -s SIGINT $pid
 }
+# Use SIGTERM or TERM, does not seem to make any difference.
 trap _term SIGTERM
 
 ros-with-env ros2 launch rplidar_ros2 sensors_launch.py &
 child=$!
 
 echo "Waiting for pid $child"
+# * Calling "wait" will then wait for the job with the specified by $child to finish, or for any signals to be fired.
+#   Due to "or for any signals to be fired", "wait" will also handle SIGTERM and it will shutdown before
+#   the node ends gracefully.
+#   The solution is to add a second "wait" call and remove the trap between the two calls.
+# * Do not use -e flag in the first wait call because wait will exit with error after catching SIGTERM.
+set +e
+wait $child
+set -e
+trap - TERM
 wait $child
 RESULT=$?
 

